src/column/writer.rs

```diff
@@ -0,0 +1,847 @@
+    // At this point we know that we need to fall back.
+    self.write_dictionary_page()?;
+    self.flush_data_pages()?;
+    self.dict_encoder = None;
+    Ok(())
+  }
+
+  /// Adds data page.
+  /// Data page is either buffered in case of dictionary encoding or written directly.
+  fn add_data_page(&mut self) -> Result<()> {
+    // Extract encoded values
+    let values = match self.dict_encoder {
+      Some(ref mut encoder) => encoder.write_indices()?,
+      None => self.encoder.flush_buffer()?
+    };
+
+    // Encode definition levels (always RLE).
+    let def_levels = if self.descr.max_def_level() > 0 {
+      self.encode_levels(
+        Encoding::RLE,
+        &self.def_levels_sink[..],
+        self.descr.max_def_level()
+      )?
+    } else {
+      vec![]
+    };
+
+    // Encode repetition levels (always RLE).
+    let mut rep_levels = if self.descr.max_rep_level() > 0 {
+      self.encode_levels(
+        Encoding::RLE,
+        &self.rep_levels_sink[..],
+        self.descr.max_rep_level()
+      )?
+    } else {
+      vec![]
+    };
+
+    // Encoded length of repetition levels.
+    let rep_levels_byte_len = rep_levels.len();
+    // Encoded length of definition levels.
+    let def_levels_byte_len = def_levels.len();
+    // Uncompressed length in bytes, compressed length depends on writer version.
+    let uncompressed_size = rep_levels_byte_len + def_levels_byte_len + values.len();
+    // Select encoding based on current encoder and writer version (v1 or v2).
+    let encoding = if self.dict_encoder.is_some() {
+      self.props.dictionary_data_page_encoding()
+    } else {
+      self.encoder.encoding()
+    };
+
+    let data_page = if self.props.writer_version() == WriterVersion::PARQUET_1_0 {
+      // Process as data page v1.
+      rep_levels.extend_from_slice(&def_levels[..]);
+      rep_levels.extend_from_slice(values.data());
+      let buffer = match self.compressor {
+        Some(ref mut cmpr) => {
+          // TODO: figure out capacity for the compressed buffer
+          let mut compressed_buf = Vec::with_capacity(rep_levels.len() / 2);
+          cmpr.compress(&rep_levels[..], &mut compressed_buf)?;
+          compressed_buf
+        }
+        None => rep_levels
+      };
+
+      Page::DataPage {
+        buf: ByteBufferPtr::new(buffer),
+        num_values: self.num_buffered_values,
+        encoding: encoding,
+        def_level_encoding: Encoding::RLE,
+        rep_level_encoding: Encoding::RLE,
+        // TODO: process statistics
+        statistics: None
+      }
+    } else {
+      // Process as data page v2.
+      let mut compressed_values = Vec::with_capacity(values.len());
+      match self.compressor {
+        Some(ref mut cmpr) => cmpr.compress(values.data(), &mut compressed_values)?,
+        None => compressed_values.extend_from_slice(values.data())
+      }
+      rep_levels.extend_from_slice(&def_levels[..]);
+      rep_levels.extend_from_slice(&compressed_values[..]);
```

Comment:
Wonder if we should keep the value buffer in a separate field in `DataPageV2`. Currently we are doing one extra copying in the write path (via `extend_from_slice`) and one extra copy in the read path (`append` in `get_next_page`).